jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java

```diff
@@ -0,0 +1,210 @@
+
+import org.jgrapht.*;
+import org.jgrapht.alg.cycle.*;
+
+/**
+ * A builder for a nice decomposition for chordal graphs. See {@link NiceDecompositionBuilder} for
+ * an explanation of nice decomposition.
+ * <p>
+ * This builder uses the perfect elimination order from {@link ChordalityInspector} to iterate over
+ * the graph. For every node it generates a node for the predecessors of the current node according
+ * to the perfect elimination order and builds a path to such a node from the node where the
+ * greatest predecessor was introduced.
+ * <p>
+ * The complexity of this algorithm is in $\mathcal{O}(|V|(|V|+|E|))$.<br>
+ * Consider the every node in the nice tree decomposition: There are exactly $|V|$ many forget
+ * nodes. There are at most $2|V|$ additionally nodes because of a join nodes. Every join node
+ * creates one additional path from root to a leaf, every such path can contain for every vertex at
+ * most one introduce node, which yields $|V|^2$ introduce nodes. Now considering the bags of the
+ * introduce nodes. On a path from root to a leaf we have at most one introduce node for every
+ * introduced vertex. Since this introduced vertex is part of the clique of the least ancestor
+ * forget node, the corresponding bag of the introduce node is smaller than the set of neighbors of
+ * this vertex. Thus the time complexity is in $\mathcal{O}(|V|(|V|+|E|))$.
+ * <p>
+ * This is a non-recursive adaption for nice tree decomposition of algorithm 2 from here: <br>
+ * Hans L. Bodlaender, Arie M.C.A. Koster, Treewidth computations I. Upper bounds, Information and
+ * Computation, Volume 208, Issue 3, 2010, Pages 259-275,
+ * 
+ * @author Ira Justus Fesefeldt (PhoenixIra)
+ * @author Timofey Chudakov
+ *
+ * @since June 2018
+ *
+ * @param <V> the vertex type of the graph
+ * @param <E> the edge type of the graph
+ */
+public class ChordalNiceDecompositionBuilder<V, E>
+    extends
+    NiceDecompositionBuilder<V>
+{
+    // the chordal graph
+    private Graph<V, E> graph;
+
+    // the perfect elimination order of graph
+    private List<V> perfectOrder;
+
+    // another representation of the perfect elimination order of graph
+    private Map<V, Integer> vertexInOrder;
+
+    /**
+     * Constructor for the nice decomposition builder of chordal graphs. The constructor
+     * already computes the decomposition.
+     * 
+     * @throws IllegalArgumentException if the graph is not chordal.
+     * @param graph the chordal graph for which a decomposition should be created
+     */
+    public ChordalNiceDecompositionBuilder(Graph<V, E> graph)
+    {
+        super();
+        ChordalityInspector<V, E> inspec = new ChordalityInspector<>(graph);
+        if (!inspec.isChordal())
+            throw new IllegalArgumentException("The given graph is not chordal.");
+        this.graph = graph;
+        this.perfectOrder = inspec.getPerfectEliminationOrder();
+        vertexInOrder = getVertexInOrder();
+        computeNiceDecomposition();
```

Comment:
Should we run the algorithm at the construction time or after the first call to some getter method? @jkinable @jsichi @d-michail what do you think?

Reply:
Since a decomposition of a graph and a tree decomposition of a graph are different things, it’s better to name it ChordalNiceTreeDecomposition and NiceTreeDecompositionBase

Reply:
Good point.